soepy/python/pre_processing/model_processing.py

```diff
@@ -13,154 +18,233 @@ def read_init_file(init_file_name):
 
-    init_dict_flat = dict()
+    # Extract the values of the type shares
+    try:
+        type_shares_non_baseline = [
+            _ for k, _ in model_params_dict["shares"].items() if "share" in k
+        ]
 
-    for group in groups:
+        num_types = len(type_shares_non_baseline) + 1
 
-        keys_ = list(init_dict[group].keys())
-        values_ = list(init_dict[group].values())
+        # Aggregate type shares in list object
+        # Share of baseline types equal to one minus sum of remaining type shares
+        type_shares = [1 - sum(type_shares_non_baseline)] + type_shares_non_baseline
 
-        for k_, key_ in enumerate(keys_):
+    except KeyError:
 
-            init_dict_flat[key_] = values_[k_]
+        type_shares = [1]
+        num_types = 1
 
-    return init_dict_flat
+    # Append derived attributes to init_dict
+    model_params_dict["derived_attr"] = {
+        "shocks_cov": shocks_cov,
+        "type_shares": type_shares,
+        "num_types": num_types,
+    }
 
+    return model_params_dict
 
-def group_parameters(init_dict, init_dict_flat):
+
+def group_parameters(model_params_dict_expanded):
     """Groups the parameters to be estimates
     in flat dictionary structure"""
 
-    init_dict_flat["gamma_0s"] = (
-        init_dict["PARAMETERS"]["gamma_0s1"],
-        init_dict["PARAMETERS"]["gamma_0s2"],
-        init_dict["PARAMETERS"]["gamma_0s3"],
+    model_params_dict_flat = {}
+
+    model_params_dict_flat["gamma_0s"] = list(
+        model_params_dict_expanded["const_wage_eq"].values()
     )
 
-    init_dict_flat["gamma_1s"] = (
-        init_dict["PARAMETERS"]["gamma_1s1"],
-        init_dict["PARAMETERS"]["gamma_1s2"],
-        init_dict["PARAMETERS"]["gamma_1s3"],
+    model_params_dict_flat["gamma_1s"] = list(
+        model_params_dict_expanded["exp_returns"].values()
     )
 
-    init_dict_flat["g_s"] = (
-        init_dict["PARAMETERS"]["g_s1"],
-        init_dict["PARAMETERS"]["g_s2"],
-        init_dict["PARAMETERS"]["g_s3"],
+    model_params_dict_flat["g_s"] = list(
+        model_params_dict_expanded["exp_accm"].values()
     )
 
-    init_dict_flat["delta_s"] = (
-        init_dict["PARAMETERS"]["delta_s1"],
-        init_dict["PARAMETERS"]["delta_s2"],
-        init_dict["PARAMETERS"]["delta_s3"],
+    model_params_dict_flat["delta_s"] = list(
+        model_params_dict_expanded["exp_deprec"].values()
     )
 
-    if init_dict["DERIVED_ATTR"]["num_types"] > 1:
+    model_params_dict_flat["const_p"] = model_params_dict_expanded["disutil_work"][
+        "const_p"
+    ]
+    model_params_dict_flat["const_f"] = model_params_dict_expanded["disutil_work"][
+        "const_f"
+    ]
+
+    model_params_dict_flat["shocks_cov"] = model_params_dict_expanded["derived_attr"][
+        "shocks_cov"
+    ]
+    model_params_dict_flat["type_shares"] = model_params_dict_expanded["derived_attr"][
+        "type_shares"
+    ]
```

Comment:
Please see if you can use loops here as the second key in `model_params_dict_expanded` is the same as the first in `model_params_dict_flat`.